jax/experimental/slab/djax.py

```diff
@@ -0,0 +1,165 @@
+# Copyright 2024 The JAX Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+from functools import partial
+import collections
+import sys
+
+import numpy as np
+import jax
+import jax.numpy as jnp
+from jax import lax
+
+from jax._src import core
+from jax._src import util
+
+import jax.experimental.slab as sl
+
+jax.config.update('jax_dynamic_shapes', True)
```

Comment:
Should we really have this at the top level of a file in JAX?

Reply:
I was using the context manager in an earlier iteration and @mattjj asked that we switch to this. Either way, turning it on ought to be fully backwards compatible in principle (right?).

Reply:
Yes it should be fully backward compatible _except_ for some edge-case errors may have to be raised later than they currently are (e.g. at lowering time rather than trace time, though TBH I forget the example here).

If we factor this to be more like a module and less like a script then it would make sense to have the jax_dynamic_shapes=True setting in the test file and in the user code, rather than in the module to be imported. That is, I agree we should remove this line from any modules-to-be-imported-by-users, and I think it'll naturally get pulled into any example files / tests instead, just like the other script-like lines currently in this file.

Reply:
We could also go back to enabling it on the inside with dynamic scope? Our use of dynamic jaxpr staging is localized to roughly one line of code.

Tests would also have to use some kind of local/dynamic scoping, otherwise they'd be turning it on for everything in process.

Reply:
I brought back configuration under a dynamic scope.

It seems that enabling dynamic shapes globally (as a side effect of import) caused failures elsewhere in the test suite in CI. I didn't dig deeper, but the issue disappeared once I implemented things under a dynamic scope and removed the global config update.